### ISSUE TITLE
Support for shape inference on several functions

### ISSUE THREAD

**Issue by tomiock** [tomiock/macrograd#12]
Implementations of:
- [ ] softmax
- [ ] relu
- [ ] MSE
- [ ] sigmoid
- [ ] cross_entropy
